Increase prewarmed WebProcess limit when Site Isolation is enabled

bnham opened the pull request:
#### ea593cc63b71d84ada69a6e9f50275bfe3044fd0
<pre>
Increase prewarmed WebProcess limit when Site Isolation is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=299494">https://bugs.webkit.org/show_bug.cgi?id=299494</a>
<a href="https://rdar.apple.com/161272150">rdar://161272150</a>

Reviewed by Sihui Liu.

When Site Isolation is enabled, we&apos;re seeing that some page loads are stalling waiting on a
WebProcess launch. To help with this, we now keep more than one prewarmed WebProcess per process
pool when Site Isolation is enabled. From initial A/B tests I&apos;ve chosen a conservative limit of 2
prewarmed processes, but I&apos;ve made this overridable via user defaults to make further tuning easier.
(For reference, Firefox uses 3 prewarmed processes when Site Isolation is enabled.)

Another change is when to trigger prewarming. Previously, we triggered prewarming when the main
frame began a provisional load (on devices with more cores) or when the main frame finished loading
(on other devices). Prewarming only when the main frame starts or finishes loading isn&apos;t sufficient
if the site has many OOP iframes. Instead, we now provide the prewarming hint to the process pool
any time any frame finishes loading.

In local testing, this seems to fix most of the cold PLT regression from enabling Site Isolation.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessPreWarming.mm
* Source/WTF/wtf/SystemTracing.h:
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.cpp:
(API::ProcessPoolConfiguration::copy):
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _prewarmedProcessIdentifiersForTesting]):
(-[WKProcessPool _prewarmedProcessIdentifier]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.mm:
(-[_WKProcessPoolConfiguration prewarmedProcessCountLimitForTesting]):
(-[_WKProcessPoolConfiguration setPrewarmedProcessCountLimitForTesting:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::notifyProcessPoolToPrewarm):
(WebKit::WebPageProxy::didFinishLoadForFrame):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcessPool::tryTakePrewarmedProcess):
(WebKit::WebProcessPool::initializeNewWebProcess):
(WebKit::WebProcessPool::prewarmProcess):
(WebKit::WebProcessPool::processDidFinishLaunching):
(WebKit::WebProcessPool::disconnectProcess):
(WebKit::WebProcessPool::processForSite):
(WebKit::WebProcessPool::didReachGoodTimeToPrewarm):
(WebKit::WebProcessPool::handleMemoryPressureWarning):
(WebKit::WebProcessPool::prewarmedProcessIdentifiers):
(WebKit::WebProcessPool::prewarmedProcessCountLimit const):
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prewarmedProcessID): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessPreWarming.mm:
(runAutomaticProcessWarmingTest):
(TEST(WKProcessPool, AutomaticProcessWarming)):
(TEST(WKProcessPool, AutomaticProcessWarmingWithMultipleProcesses)):
(TEST(WKProcessPool, PrewarmedProcessCrash)):
(TEST(WKProcessPool, TryUsingPrewarmedProcessThatJustCrashed)):
(TEST(WebKit, TryUsingPrewarmedProcessThatJustCrashed)): Deleted.

Canonical link: <a href="https://commits.webkit.org/300594@main">https://commits.webkit.org/300594@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b77190120b6e9a40c2cf7aa3186dce709a2246ec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/123184 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/131/builds/42898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/33595 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/129891 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/75296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/56a1ab1a-c338-4447-9b60-9ea41cb4bdb9) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/125061 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/43622 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/123/builds/51493 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/93644 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/75296 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/62c10c13-b61b-4273-ad54-f8683e2a499f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/126135 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/34765 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/110245 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/74274 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/206ba84b-06a1-4113-af22-99ef14424375) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/133/builds/33734 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/28401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/73407 "Built successfully") | | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/115343 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/104477 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/136/builds/28627 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/132606 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/121715 "Built successfully and passed tests") | [  ~~🛠 vision~~](https://ews-build.webkit.org/#/builders/128/builds/50134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/123/builds/51493 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/102133 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [  ~~🛠 vision-sim~~](https://ews-build.webkit.org/#/builders/121/builds/50510 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/106466 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/101989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25922 "Built successfully and passed tests") | [⏳ 🧪 vision-wk2 ](https://ews-build.webkit.org/#/builders/visionOS-2-Simulator-WK2-Tests-EWS "Waiting in queue, processing has not started yet") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/25569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/46908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/127/builds/49989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/55750 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/151980 "Built successfully") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/49459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/38864 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/52809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/51138 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/WebProcessPool.cpp:
```diff
@@ -2053,10 +2084,45 @@ void WebProcessPool::removeProcessFromOriginCacheSet(WebProcessProxy& process)
+    Ref pagePreferences = page.preferences();
+    bool siteIsolationEnabled = pagePreferences->siteIsolationEnabled();
+
+    if (pagePreferences->siteIsolationSharedProcessEnabled())
```
Should be `siteIsolationEnabled && pagePreferences->siteIsolationSharedProcessEnabled()` (as we currently don't set `siteIsolationSharedProcessEnabled` based on `siteIsolationEnabled`)

bnham replied:
Sounds good, I'm just going to nuke this altogether and use a default limit of 2 when SI is enabled (whether or not shared process is enabled).

bnham committed:
Increase prewarmed WebProcess limit when Site Isolation is enabled
https://bugs.webkit.org/show_bug.cgi?id=299494
rdar://161272150

Reviewed by Sihui Liu.

When Site Isolation is enabled, we're seeing that some page loads are stalling waiting on a
WebProcess launch. To help with this, we now keep more than one prewarmed WebProcess per process
pool when Site Isolation is enabled. From initial A/B tests I've chosen a conservative limit of 2
prewarmed processes, but I've made this overridable via user defaults to make further tuning easier.
(For reference, Firefox uses 3 prewarmed processes when Site Isolation is enabled.)

Another change is when to trigger prewarming. Previously, we triggered prewarming when the main
frame began a provisional load (on devices with more cores) or when the main frame finished loading
(on other devices). Prewarming only when the main frame starts or finishes loading isn't sufficient
if the site has many OOP iframes. Instead, we now provide the prewarming hint to the process pool
any time any frame finishes loading.

In local testing, this seems to fix most of the cold PLT regression from enabling Site Isolation.

Test: Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessPreWarming.mm
* Source/WTF/wtf/SystemTracing.h:
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.cpp:
(API::ProcessPoolConfiguration::copy):
* Source/WebKit/UIProcess/API/APIProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPool.mm:
(-[WKProcessPool _prewarmedProcessIdentifiersForTesting]):
(-[WKProcessPool _prewarmedProcessIdentifier]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/WKProcessPoolPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKProcessPoolConfiguration.mm:
(-[_WKProcessPoolConfiguration prewarmedProcessCountLimitForTesting]):
(-[_WKProcessPoolConfiguration setPrewarmedProcessCountLimitForTesting:]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::notifyProcessPoolToPrewarm):
(WebKit::WebPageProxy::didFinishLoadForFrame):
* Source/WebKit/UIProcess/WebProcessPool.cpp:
(WebKit::WebProcessPool::establishRemoteWorkerContextConnectionToNetworkProcess):
(WebKit::WebProcessPool::tryTakePrewarmedProcess):
(WebKit::WebProcessPool::initializeNewWebProcess):
(WebKit::WebProcessPool::prewarmProcess):
(WebKit::WebProcessPool::processDidFinishLaunching):
(WebKit::WebProcessPool::disconnectProcess):
(WebKit::WebProcessPool::processForSite):
(WebKit::WebProcessPool::didReachGoodTimeToPrewarm):
(WebKit::WebProcessPool::handleMemoryPressureWarning):
(WebKit::WebProcessPool::prewarmedProcessIdentifiers):
(WebKit::WebProcessPool::prewarmedProcessCountLimit const):
(WebKit::WebProcessPool::processForNavigation):
(WebKit::WebProcessPool::prewarmedProcessID): Deleted.
* Source/WebKit/UIProcess/WebProcessPool.h:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/ProcessPreWarming.mm:
(runAutomaticProcessWarmingTest):
(TEST(WKProcessPool, AutomaticProcessWarming)):
(TEST(WKProcessPool, AutomaticProcessWarmingWithMultipleProcesses)):
(TEST(WKProcessPool, PrewarmedProcessCrash)):
(TEST(WKProcessPool, TryUsingPrewarmedProcessThatJustCrashed)):
(TEST(WebKit, TryUsingPrewarmedProcessThatJustCrashed)): Deleted.

Canonical link: https://commits.webkit.org/300594@main
